raft.go

```diff
@@ -308,7 +315,54 @@ func (r *Raft) runCandidate() {
 		case rpc := <-r.rpcCh:
 			r.mainThreadSaturation.working()
 			r.processRPC(rpc)
+		case vote := <-prevoteCh:
+			// This a pre-vote case but could lead to winning an election, in the case that majority of other nodes
+			// don't support pre-vote or have pre-vote deactivated.
+			r.mainThreadSaturation.working()
+			r.logger.Debug("got a prevote!!", "from", vote.voterID, "term", vote.Term, "tally", grantedVotes)
+			// Check if the term is greater than ours, bail
+			if vote.Term > r.getCurrentTerm() && !vote.PreVote {
+				r.logger.Debug("newer term discovered, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+			if vote.Term > term && vote.PreVote {
+				r.logger.Debug("newer term discovered on pre-vote, fallback to follower", "term", vote.Term)
+				r.setState(Follower)
+				r.setCurrentTerm(vote.Term)
+				return
+			}
+
+			// Check if the vote is granted
+			if vote.Granted {
+				if !vote.PreVote {
+					grantedVotes++
+					r.logger.Debug("vote granted", "from", vote.voterID, "term", vote.Term, "tally", grantedVotes)
+				} else {
+					preVoteGrantedVotes++
+					r.logger.Debug("prevote granted", "from", vote.voterID, "term", vote.Term, "tally", preVoteGrantedVotes)
+				}
+			}
+
+			// Check if we've become the leader
+			// If we won the election because the majority of nodes don't support pre-vote (or pre-vote is deactivated)
+			// set our state to leader and our term to the pre-vote term.
+			if grantedVotes >= votesNeeded {
+				r.logger.Info("election won", "term", vote.Term, "tally", grantedVotes)
+				r.setState(Leader)
+				r.setLeader(r.localAddr, r.localID)
 
+				r.setCurrentTerm(term)
```

Comment:
This is the part that I think we should probably also `persistVote` for ourselves even though in this code you didn't count our own preVote in the quorum. I don't think this is necessarily incorrect as it is, but I wonder if we'll violate assumptions later that we are leader in a term but we never voted in that term? I think it's safe in all cases to decide to "upgrade" our own preVote to a real vote here anyway and avoids ever needing to think about that case.

Technically if we keep this approach we could get to this point even faster by checking for `grantedVotes >= (votesNeeded-1)` since we can always upgrade our own vote to make up the quorum if enough other peers voted for real.

Reply:
Actually I came up with [a scenario](https://github.com/hashicorp/raft/pull/530#discussion_r1019159636) where the cluster could get stuck with one node running PreVote campaigns that never get enough support to complete but prevent other nodes getting elected despite a quorum of healthy nodes with this logic. If we change it to only need quorum - 1 and upgrade our own vote then those cases can't happen...